### PR TITLE
Upgrade to go1.9

### LINKED ISSUE #49484: Update to Go 1.9
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

This is a followup on https://github.com/kubernetes/kubernetes/issues/38228, but for Go 1.9

We should do the same kind of testing now to possibly catch bugs in either Go or Kubernetes when doing this kind of dependency version bump.

Ideally we'd have automated tests against Go tip (https://github.com/kubernetes/test-infra/issues/1399), but we don't have that at the moment, so we have to do it kind of manually.

We should also decide which k8s version we should target. k8s v1.8 or v1.9?

What Go versions have been used so far?
 - Kubernetes v1.7.0 - go1.8.3 [ref](https://github.com/kubernetes/kubernetes/blob/v1.7.0/build/build-image/cross/Dockerfile)
 - Kubernetes v1.6.0 - go1.7.5 [ref](https://github.com/kubernetes/kubernetes/blob/v1.6.0/build/build-image/cross/Dockerfile)
 - Kubernetes v1.5.0 - go1.7.4 [ref](https://github.com/kubernetes/kubernetes/blob/v1.5.0/build-tools/build-image/cross/Dockerfile)
 - Kubernetes v1.4.0 - go1.6.3 [ref](https://github.com/kubernetes/kubernetes/blob/v1.4.0/build/build-image/cross/Dockerfile)
 - Kubernetes v1.3.0 - go1.6.2 [ref](https://github.com/kubernetes/kubernetes/blob/v1.3.0/build/build-image/cross/Dockerfile)
 - Kubernetes v1.2.0 - go1.4.2 [ref](https://github.com/kubernetes/kubernetes/blob/v1.2.0/build/build-image/cross/Dockerfile)

Seems like the trend has been to stay on the same Go version for two releases and then bump.
That'd mean we'd switch to Go1.9 in k8s v1.9

Everyone happy with that? Or is there any reason for us to go start using go1.9 in k8s v1.8?

In any case, we should indeed start testing go1.9 out now...

cc @kubernetes/sig-cluster-lifecycle-feature-requests @kubernetes/sig-testing-feature-requests @kubernetes/sig-release-feature-requests @bradfitz @jessfraz @kubernetes/sig-scalability-feature-requests

### PR DESCRIPTION
**Pull request by cblecker**
**What this PR does / why we need it**:
Upgrade to go1.9. Upgrading is good. It's "the best golang release ever"!

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #49484

**Special notes for your reviewer**:

**Release note**:

```release-note
Upgrade to go1.9
```

/assign @luxas @ixdy @wojtek-t

### CODE REVIEW

**Comment by cblecker** [comment]
Probably needs https://github.com/bazelbuild/rules_go/issues/747 to get go1.9 bazel rules

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -334,7 +334,7 @@ EOF
   local go_version
   go_version=($(go version))
   local minimum_go_version
-  minimum_go_version=go1.8.3
+  minimum_go_version=go1.9
```
Comment (xiangpengzhao) [inline hack/lib/golang.sh]: If there are no compatibility issues (e.g., K8S import some packages only in go1.9), maybe we'd better keep `minimum_go_version` to be the last version we used. @cblecker WDYT?

**Comment by gtirloni** [comment]
Cannot compile Kubernetes with go 1.9 on Linux:

```
$ make
+++ [0826 13:29:12] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
    k8s.io/kubernetes/vendor/github.com/jteeuwen/go-bindata/go-bindata
# runtime
/usr/local/go/src/runtime/mstkbar.go:151:10: debug.gcstackbarrieroff undefined (type struct { allocfreetrace int32; cgocheck int32; efence int32; gccheckmark int32; gcpacertrace int32; gcshrinkstackoff int32; gcrescanstacks int32; gcstoptheworld int32; gctrace int32; invalidptr int32; sbrk int32; scavenge int32; scheddetail int32; schedtrace int32 } has no field or method gcstackbarrieroff)
/usr/local/go/src/runtime/mstkbar.go:162:24: division by zero
/usr/local/go/src/runtime/mstkbar.go:162:43: invalid expression unsafe.Sizeof(composite literal)
/usr/local/go/src/runtime/mstkbar.go:162:44: undefined: stkbar
/usr/local/go/src/runtime/mstkbar.go:212:4: gp.stkbar undefined (type *g has no field or method stkbar)
/usr/local/go/src/runtime/mstkbar.go:213:15: gp.stkbar undefined (type *g has no field or method stkbar)
/usr/local/go/src/runtime/mstkbar.go:216:23: undefined: stackBarrierPC
/usr/local/go/src/runtime/mstkbar.go:226:28: gp.stkbarPos undefined (type *g has no field or method stkbarPos)
/usr/local/go/src/runtime/mstkbar.go:227:19: gp.stkbarPos undefined (type *g has no field or method stkbarPos)
/usr/local/go/src/runtime/mstkbar.go:248:41: undefined: stkbar
/usr/local/go/src/runtime/mstkbar.go:227:19: too many errors
!!! [0826 13:29:12] Call tree:
!!! [0826 13:29:12]  1: /home/gtirloni/go/src/k8s.io/kubernetes/hack/lib/golang.sh:679 kube::golang::build_kube_toolchain(...)
!!! [0826 13:29:12]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0826 13:29:12] Call tree:
!!! [0826 13:29:12]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [Makefile.generated_files:302: _output/bin/deepcopy-gen] Error 1
make: *** [Makefile:481: generated_files] Error 2
```

I merged master into this PR but got the same results.

**Comment by wojtek-t** [comment]
@kubernetes/sig-scalability-misc  FYI

File: hack/lib/golang.sh (hunk 1 above)
Comment (cblecker) [reply hack/lib/golang.sh]: Yeah, you're right, we don't need to bump the minimum just yet.

**Comment by cblecker** [comment]
@ixdy This should be good to push kube-cross and re-test :)

File: build/root/WORKSPACE [hunk 2]
```diff
@@ -43,7 +43,7 @@ load("@io_bazel_rules_go//go:def.bzl", "go_repositories")
 load("@io_bazel_rules_docker//docker:docker.bzl", "docker_repositories", "docker_pull")
 
 go_repositories(
-    go_version = "1.8.3",
+    go_version = "1.9",
```
Comment (ixdy) [inline build/root/WORKSPACE]: this will not work until https://github.com/bazelbuild/rules_go/pull/753 is merged and the rules_go dependency is updated.

**Comment by cblecker** [comment]
/test pull-kubernetes-bazel
Looks like https://github.com/bazelbuild/rules_go/pull/753 was merged

**Comment by ixdy** [comment]
@cblecker you will need to update [build/root/WORKSPACE](https://github.com/kubernetes/kubernetes/blob/master/build/root/WORKSPACE#L1-L6) and probably `hack/update-bazel.sh` and `hack/verify-bazel.sh` too.

**Comment by cblecker** [comment]
```
W0828 21:28:20.295] java.lang.RuntimeException: Unrecoverable error while evaluating node 'SKYLARK_IMPORTS_LOOKUP:@io_bazel_rules_go//go/private:library.bzl (in workspace)' (requested by nodes 'SKYLARK_IMPORTS_LOOKUP:@io_bazel_rules_go//go/private:wrappers.bzl (in workspace)', 'SKYLARK_IMPORTS_LOOKUP:@io_bazel_rules_go//go/private:test.bzl (in workspace)', 'SKYLARK_IMPORTS_LOOKUP:@io_bazel_rules_go//go/private:cgo.bzl (in workspace)', 'SKYLARK_IMPORTS_LOOKUP:@io_bazel_rules_go//go/private:binary.bzl (in workspace)')
```
:(

**Comment by ixdy** [comment]
yeah, I think we need to bump the version of bazel we're running in CI - 0.5.2 is probably too old now.

in any case, it looks like the new go rules break some of our custom rules (which I expected), which will require additional fixes:
```console
$ make bazel-build 
ERROR: [bazel-cache]/external/io_kubernetes_build/defs/go.bzl:1:1: file '@io_bazel_rules_go//go:def.bzl' does not contain symbol 'GoSource'
ERROR: [bazel-cache]/external/io_kubernetes_build/defs/go.bzl:2:1: file '@io_bazel_rules_go//go/private:common.bzl' does not contain symbol 'get_go_toolchain'
ERROR: error loading package 'pkg/generated/openapi': Extension file 'defs/go.bzl' has errors
ERROR: error loading package 'pkg/generated/openapi': Extension file 'defs/go.bzl' has errors
INFO: Elapsed time: 2.575s
```

I can follow up on those fixes, but it probably won't happen until later this week.

In all honesty, I don't expect to get go1.9 in before the k8s1.8 code freeze. It seems like a pretty large change this late in the cycle.

**Comment by cblecker** [comment]
Yeah, I think from [the issue](https://github.com/kubernetes/kubernetes/issues/49484) the consensus was that we'll do this after 1.8. So no rush on any of this from my point of view.

**Comment by ixdy** [comment]
testing with bazel 0.5.4 here: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/51375/pull-kubernetes-bazel/38913/

**Comment by ixdy** [comment]
right, that's still broken.

**Comment by dims** [comment]
Hey @cblecker, wanna rebase so we can see if there's any issues with say the unit tests?

**Comment by cblecker** [comment]
@dims I think we're still waiting on bazel 0.6.0: https://github.com/kubernetes/kubernetes/issues/49484#issuecomment-331586783

**Comment by luxas** [comment]
ETA on 0.6.0?

**Comment by ixdy** [comment]
Bazel 0.6.0 is out. We're now trying to figure how how best to support multiple versions of Bazel in prow.

**Comment by luxas** [comment]
@ixdy with multiarch support or not?

**Comment by wojtek-t** [comment]
@cblecker @ixdy - what is blocking this PR now? If we want this in 1.9 timeframe, this is the best time to do it as early part of release cycle.

**Comment by ixdy** [comment]
@wojtek-t I just sent out a PR to update CI to use bazel 0.6.1. That clears the way for us to bump `rules_go`, which then lets us use go1.9+. Hoping to get all of that to land by end of week.

**Comment by wojtek-t** [comment]
> @wojtek-t I just sent out a PR to update CI to use bazel 0.6.1. That clears the way for us to bump rules_go, which then lets us use go1.9+. Hoping to get all of that to land by end of week.

awesome!

**Comment by dims** [comment]
/retest

Looks like that has merged, should a retest work now? Let's see!

**Comment by luxas** [comment]
@cblecker @ixdy Maybe push the new images so we can examine the results?
Or does this need a rebase?

**Comment by cblecker** [comment]
I think `gcr.io/google_containers/kube-cross:v1.9.1-1` needs to be built and pushed

**Comment by luxas** [comment]
> I think gcr.io/google_containers/kube-cross:v1.9.1-1 needs to be built and pushed

Yes, can you do that @cblecker?

**Comment by cblecker** [comment]
I cannot as I am not a googler 😄 
Can we get a helping hand when you have time @ixdy ?

**Comment by luxas** [comment]
> I cannot as I am not a googler :smile:

Heh, me neither ;)

**Comment by ixdy** [comment]
Once #53839 merges we should be able to bump to go 1.9.1 in the bazel build. I'll work on building cross-build images tomorrow, too.

**Comment by ixdy** [comment]
That comment ^ means #53839 merged. Please rebase and I'll push new images.

**Comment by cblecker** [comment]
@ixdy rebased! back over to you :)

**Comment by ixdy** [comment]
`gcr.io/google_containers/kube-cross:v1.9.1-1` has been pushed.

/retest

**Comment by ixdy** [comment]
The unit test failure (in bazel-test) might be legitimate. It's not failing on the `pull-kubernetes-unit` job because that job is still using go1.8.3.

**Comment by cblecker** [comment]
Yup, looks legitimate when I run locally:
```
$ go test -v ./plugin/pkg/admission/webhook
=== RUN   TestAdmit
2017/10/16 15:11:08 http: TLS handshake error from 127.0.0.1:62299: remote error: tls: bad certificate
E1016 15:11:08.664921   90108 admission.go:199] failed calling admission webhook "disallowReason": Post https://127.0.0.1:62298/disallowReason: x509: certificate is not valid for any names, but wanted to match webhook-test.default.svc
2017/10/16 15:11:08 http: TLS handshake error from 127.0.0.1:62302: remote error: tls: bad certificate
E1016 15:11:08.670346   90108 admission.go:199] failed calling admission webhook "internalErr C": Post https://127.0.0.1:62298/internalErr: x509: certificate is not valid for any names, but wanted to match webhook-test.default.svc
2017/10/16 15:11:08 http: TLS handshake error from 127.0.0.1:62300: remote error: tls: bad certificate
E1016 15:11:08.670615   90108 admission.go:199] failed calling admission webhook "internalErr B": Post https://127.0.0.1:62298/internalErr: x509: certificate is not valid for any names, but wanted to match webhook-test.default.svc
2017/10/16 15:11:08 http: TLS handshake error from 127.0.0.1:62301: remote error: tls: bad certificate
E1016 15:11:08.670893   90108 admission.go:199] failed calling admission webhook "internalErr A": Post https://127.0.0.1:62298/internalErr: x509: certificate is not valid for any names, but wanted to match webhook-test.default.svc
2017/10/16 15:11:08 http: TLS handshake error from 127.0.0.1:62303: remote error: tls: bad certificate
E1016 15:11:08.676384   90108 admission.go:199] failed calling admission webhook "allow": Post https://127.0.0.1:62298/allow: x509: certificate is not valid for any names, but wanted to match webhook-test.default.svc
2017/10/16 15:11:08 http: TLS handshake error from 127.0.0.1:62304: remote error: tls: bad certificate
E1016 15:11:08.679835   90108 admission.go:199] failed calling admission webhook "disallow": Post https://127.0.0.1:62298/disallow: x509: certificate is not valid for any names, but wanted to match webhook-test.default.svc
--- FAIL: TestAdmit (0.02s)
        admission_test.go:221: "match & disallow ii": expected allowed=false, but got err=<nil>
        admission_test.go:226: "match & disallow ii": expected an error saying "you shall not pass", but got <nil>
        admission_test.go:221: "match & disallow": expected allowed=false, but got err=<nil>
        admission_test.go:226: "match & disallow": expected an error saying "without explanation", but got <nil>
=== RUN   TestGroup
--- PASS: TestGroup (0.00s)
=== RUN   TestVersion
--- PASS: TestVersion (0.00s)
=== RUN   TestOperation
--- PASS: TestOperation (0.00s)
=== RUN   TestResource
--- PASS: TestResource (0.00s)
=== RUN   TestDefaultServiceResolver
=== RUN   TestDefaultServiceResolver/scenario_0
=== RUN   TestDefaultServiceResolver/scenario_1
=== RUN   TestDefaultServiceResolver/scenario_2
--- PASS: TestDefaultServiceResolver (0.00s)
    --- PASS: TestDefaultServiceResolver/scenario_0 (0.00s)
    --- PASS: TestDefaultServiceResolver/scenario_1 (0.00s)
    --- PASS: TestDefaultServiceResolver/scenario_2 (0.00s)
FAIL
exit status 1
FAIL    k8s.io/kubernetes/plugin/pkg/admission/webhook  0.117s
```

**Comment by cblecker** [comment]
And ran the same inside a go1.8.3 environment, and tests passed:
```
$ go test -v ./plugin/pkg/admission/webhook
=== RUN   TestAdmit
got req: /allow
got req: /disallow
got req: /disallowReason
got req: /internalErr
E1016 15:30:57.032550   19814 admission.go:199] failed calling admission webhook "internalErr C": an error on the server ("webhook internal server error") has prevented the request from succeeding
got req: /internalErr
E1016 15:30:57.035062   19814 admission.go:199] failed calling admission webhook "internalErr A": an error on the server ("webhook internal server error") has prevented the request from succeeding
got req: /internalErr
E1016 15:30:57.037306   19814 admission.go:199] failed calling admission webhook "internalErr B": an error on the server ("webhook internal server error") has prevented the request from succeeding
--- PASS: TestAdmit (0.07s)
=== RUN   TestGroup
--- PASS: TestGroup (0.00s)
=== RUN   TestVersion
--- PASS: TestVersion (0.00s)
=== RUN   TestOperation
--- PASS: TestOperation (0.00s)
=== RUN   TestResource
--- PASS: TestResource (0.00s)
=== RUN   TestDefaultServiceResolver
=== RUN   TestDefaultServiceResolver/scenario_0
=== RUN   TestDefaultServiceResolver/scenario_1
=== RUN   TestDefaultServiceResolver/scenario_2
--- PASS: TestDefaultServiceResolver (0.00s)
    --- PASS: TestDefaultServiceResolver/scenario_0 (0.00s)
    --- PASS: TestDefaultServiceResolver/scenario_1 (0.00s)
    --- PASS: TestDefaultServiceResolver/scenario_2 (0.00s)
PASS
ok      k8s.io/kubernetes/plugin/pkg/admission/webhook  0.148s
```

**Comment by cblecker** [comment]
Opened #54020 for sig-node to have a look at it.

**Comment by dims** [comment]
@cblecker - Can you please check https://github.com/kubernetes/kubernetes/pull/54030 if it works for you?

**Comment by ixdy** [comment]
LGTM. @wojtek-t should we tag this for merge?

**Comment by wojtek-t** [comment]
I think we should try.

@shyamjvs @porridge  @kubernetes/sig-scalability-pr-reviews - in case you observe any regression once this is merged.

/lgtm

**Comment by shyamjvs** [comment]
/test pull-kubernetes-kubemark-e2e-gce-big
/hold

Let's try running kubemark-500.

**Comment by shyamjvs** [comment]
/hold cancel
/lgtm

kubemark-500 passed - https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/51375/pull-kubernetes-kubemark-e2e-gce-big/30/

**Comment by dims** [comment]
Very cool

/lgtm

**Pull request merged**